File: victorious/victorious/UI/Content View/Experience Enhancers/VImageAnimationOperation.m
////
////  VAnimationImageView.m
////  victorious
////
////  Created by Vincent Ho on 1/26/16.
////  Copyright © 2016 Victorious. All rights reserved.
////
//
//#import "VImageAnimationOperation.h"
//
//@interface VAnimationOperation()
//
//@property (nonatomic, strong) UIImageView *animationImageView;
//@property (nonatomic, strong) NSTimer *animationTimer;
//@property (nonatomic) NSInteger frame;
//@property (nonatomic) dispatch_semaphore_t sem;
//
//@end
//
//
//@implementation VAnimationOperation
//
//- (instancetype)initWithFrame:(CGRect)frame
//{
//    self = [super init];
//    if (self)
//    {
//        self.animationImageView = [[UIImageView alloc] initWithFrame:frame];
//        self.animationDuration = 1;
//        self.frame = -1;
////        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(stopAnimating) name:ANIMATION_IMAGE_VIEW_MEMORY_WARNING_STOP object:nil];
//    }
//    return self;
//}
//
//- (void)setContentMode:(UIViewContentMode)contentMode
//{
//    _contentMode = contentMode;
//    self.animationImageView.contentMode = contentMode;
//}
//
//- (BOOL)isAnimating
//{
//    return self.frame != -1;
//}
//
//- (void)updateFrame:(id)sender
//{
//    self.frame++;
//    [self updateImageFrame];
//}
//
//- (void)updateImageFrame
//{
//    if (self.frame == (int)_animationSequence.count)
//    {
//        [self stopAnimating];
//    }
//    else if (self.frame >= 0)
//    {
//        self.animationImageView.image = self.animationSequence[self.frame];
//    }
//    else
//    {
//        self.animationImageView.image = nil;
//    }
//}
//
//- (void)startAnimatingOn:(UIView *)view
//{
//    if (self.animationSequence.count == 0)
//    {
//        [self stopAnimating];
//        return;
//    }
//    dispatch_async(dispatch_get_main_queue(), ^{
//        [view addSubview:self.animationImageView];
//    });
//    self.frame = 0;
//    CGFloat frameDuration = self.animationDuration/self.animationSequence.count;
//    self.animationTimer = [NSTimer timerWithTimeInterval:frameDuration target:self selector:@selector(updateFrame:) userInfo:nil repeats:YES];
//    [[NSRunLoop mainRunLoop] addTimer:self.animationTimer forMode: NSDefaultRunLoopMode];
//}
//
//- (void)startAnimatingOn:(UIView *)view withSemaphore:(dispatch_semaphore_t)sem
//{
//    _sem = sem;
//    [self startAnimatingOn:view];
//}
//
//- (void)stopAnimating
//{
//    if (self.delegate)
//    {
//        [self.delegate animation:self didFinishAnimating:(self.frame == (NSInteger)self.animationSequence.count)];
//    }
//    if (self.sem)
//    {
//        dispatch_semaphore_signal(self.sem);
//    }
//    self.frame = -1;
//    [self.animationTimer invalidate];
//    [self updateImageFrame];
//    [self.animationImageView removeFromSuperview];
//}
//
//- (void)startFlightFor:(NSTimeInterval)flightDuration on:(UIView *)view center:(CGPoint)center frame:(CGRect)frame
//{
//    self.animationImageView.center = center;
//    dispatch_async(dispatch_get_main_queue(), ^
//    {
//        [UIView animateWithDuration:flightDuration
//                              delay:0.0f
//                            options:UIViewAnimationOptionCurveLinear
//                         animations:^
//         {
//             CGFloat randomLocationX = arc4random_uniform(CGRectGetWidth(frame));
//             CGFloat randomLocationY = arc4random_uniform(CGRectGetHeight(frame));
//             self.animationImageView.center = CGPointMake(randomLocationX, randomLocationY);
//         }
//                         completion:^(BOOL finished)
//         {
//             [self startAnimatingOn:view];
//         }];
//    });
//}
//
//- (void)dealloc
//{
//    [[NSNotificationCenter defaultCenter] removeObserver:self];
//}
//
//@end

File: victorious/victorious/UI/Content View/Experience Enhancers/VImageAnimationOperation.swift
//
//  VImageAnimationOperation.swift
//  victorious
//
//  Created by Vincent Ho on 1/26/16.
//  Copyright © 2016 Victorious. All rights reserved.
//

import UIKit

@objc protocol VImageAnimationOperationDelegate {
    /// Should remove imageview after this call
    func animation(animation: VImageAnimationOperation, didFinishAnimating completed:Bool)
    
    /// Should update UIImageView
    func animation(animation: VImageAnimationOperation, updatedToImage image:UIImage?)
}

class VImageAnimationOperation: Operation {
    
    weak var delegate: VImageAnimationOperationDelegate?
    
    var animationImageView: UIImageView = UIImageView()
    var animationDuration: Float = 0
    var animationSequence = [UIImage]()
    private var currentFrame: Int = -1
    private var animationTimer: VTimerManager = VTimerManager()
    var ballisticAnimationBlock: (( ()->() )->(Void)) = { completion in
        completion()
    }
    
    func isAnimating() -> Bool {
        if completedAnimation() {
            return false
        }
        return currentFrame != -1
    }
    
    /// If nil or empty animation sequence, by default the animation is done.
    func completedAnimation() -> Bool {
        if animationSequence.isEmpty {
            return true
        }
        else {
            return currentFrame == animationSequence.count
        }
    }
    
    func updateFrame() {
        updateImageFrame()
        currentFrame++
    }
    
    func updateImageFrame() {
        if cancelled || completedAnimation() {
            stopAnimating()
        }
        else {
            let image: UIImage? = animationSequence[currentFrame]
            delegate?.animation(self, updatedToImage: image)
        }
    }
    
    func beginAnimation() {
        
        let frameDuration: Float = animationDuration/Float(animationSequence.count)
        ballisticAnimationBlock() {
            VTimerManager.addTimerManagerWithTimeInterval(NSTimeInterval(frameDuration), target: self, selector: "updateFrame", userInfo: nil, repeats: true, toRunLoop: NSRunLoop.mainRunLoop(), withRunMode: NSRunLoopCommonModes)
        }
    }
    
    func startAnimating() {
        currentFrame = 0
        if animationSequence.isEmpty {
            stopAnimating()
        }
        else {
            beginAnimation()
        }
    }
    
    func stopAnimating() {
        delegate?.animation(self, updatedToImage: nil)
        let finishedAnimation: Bool = currentFrame == animationSequence.count
        delegate?.animation(self, didFinishAnimating:finishedAnimation)
        animationTimer.invalidate()
        finishedExecuting()
    }
    
    override func start() {
        startAnimating()
        super.start()
    }
    
    override func cancel() {
        stopAnimating()
        super.cancel()
    }
}
